libfreerdp/core/license.c

```diff
@@ -132,21 +139,178 @@ static void license_print_product_info(LICENSE_PRODUCT_INFO* productInfo)
 	free(ProductId);
 }
 
-static void license_print_scope_list(SCOPE_LIST* scopeList)
+static void license_print_scope_list(const SCOPE_LIST* scopeList)
 {
 	int index;
-	LICENSE_BLOB* scope;
+	const LICENSE_BLOB* scope;
 	WLog_INFO(TAG, "ScopeList (%"PRIu32"):", scopeList->count);
 
 	for (index = 0; index < scopeList->count; index++)
 	{
 		scope = &scopeList->array[index];
-		WLog_INFO(TAG, "\t%s", (char*) scope->data);
+		WLog_INFO(TAG, "\t%s", (const char*) scope->data);
 	}
 }
-
 #endif
 
+
+static const char licenseStore[] = "licenses";
+
+static BOOL computeCalHash(const char *hostname, char *hashStr)
+{
+	WINPR_DIGEST_CTX* sha1 = NULL;
+	BOOL ret = FALSE;
+	BYTE hash[20];
+	int i;
+
+	if (!(sha1 = winpr_Digest_New()))
+		goto out;
+	if (!winpr_Digest_Init(sha1, WINPR_MD_SHA1))
+		goto out;
+	if (!winpr_Digest_Update(sha1, (const BYTE *)hostname, strlen(hostname)))
+		goto out;
+	if (!winpr_Digest_Final(sha1, hash, sizeof(hash)))
+		goto out;
+
+	for (i = 0; i < sizeof(hash); i++, hashStr += 2)
+		sprintf_s(hashStr, 3, "%.2x", hash[i]);
+
+    ret = TRUE;
+out:
+	winpr_Digest_Free(sha1);
+	return ret;
+}
+
+static BOOL saveCal(rdpSettings *settings, const BYTE *data, int length, char *hostname)
+{
+	char hash[41];
+	FILE *fp;
+	char *licenseStorePath = NULL;
+	char filename[MAX_PATH], filenameNew[MAX_PATH];
+	char *filepath, *filepathNew;
+	size_t written;
+	BOOL ret = FALSE;
+
+	if (!PathFileExistsA(settings->ConfigPath))
+	{
+		if (!PathMakePathA(settings->ConfigPath, 0))
+		{
+			WLog_ERR(TAG, "error creating directory '%s'", settings->ConfigPath);
+			goto fail;
+		}
+		WLog_INFO(TAG, "creating directory %s", settings->ConfigPath);
+	}
+
+	if (!(licenseStorePath = GetCombinedPath(settings->ConfigPath, licenseStore)))
+		goto fail;
+
+	if (!PathFileExistsA(licenseStorePath))
+	{
+		if (!PathMakePathA(licenseStorePath, 0))
+		{
+			WLog_ERR(TAG, "error creating directory '%s'", licenseStorePath);
+			goto fail_create_store_path;
+		}
+		WLog_INFO(TAG, "creating directory %s", licenseStorePath);
+	}
+
+	if (!computeCalHash(hostname, hash))
+		goto fail_compute_hash;
+	sprintf_s(filename, sizeof(filename)-1, "%s.cal", hash);
+	sprintf_s(filenameNew, sizeof(filenameNew)-1, "%s.cal.new", hash);
+
+	if (!(filepath = GetCombinedPath(licenseStorePath, filename)))
+		goto fail_file_path;
+
+	if (!(filepathNew = GetCombinedPath(licenseStorePath, filenameNew)))
+		goto fail_file_path_new;
+
+	fp = fopen(filepathNew, "wb");
+	if (!fp)
+		goto fail_file_open;
+
+	written = fwrite(data, length, 1, fp);
+	fclose(fp);
+
+	if (written != 1)
+	{
+		DeleteFile(filepathNew);
+		goto fail_file_open;
+	}
+
+	ret = MoveFileEx(filepathNew, filepath, MOVEFILE_REPLACE_EXISTING);
+
+fail_file_open:
```

Comment:
initializing variables with `NULL` could avoid having more than one `goto` label

Reply:
good catch, changing this.